Mathlib/Analysis/SpecialFunctions/Arcosh.lean

```diff
@@ -0,0 +1,256 @@
+
+open Function Filter Set NNReal
+
+open scoped Topology
+
+namespace Real
+
+variable {x y : ℝ}
+
+/-- `arcosh` is defined using a logarithm, `arcosh x = log (x + √(x ^ 2 - 1))`. -/
+@[pp_nodot]
+def arcosh (x : ℝ) :=
+  log (x + √(x ^ 2 - 1))
+
+theorem exp_arcosh (x : ℝ) (hx : 1 ≤ x) : exp (arcosh x) = x + √(x ^ 2 - 1) := by
+  apply exp_log
+  positivity
+
+@[simp]
+theorem arcosh_zero : arcosh 1 = 0 := by simp [arcosh]
+
+lemma x_add_sqrt_x_sq_sub_one_inv (x : ℝ) (hx : 1 ≤ x) :
+  (x + √(x ^ 2 - 1))⁻¹ = x - √(x ^ 2 - 1) := by
+    apply inv_eq_of_mul_eq_one_right
+    rw [← pow_two_sub_pow_two, sq_sqrt (sub_nonneg_of_le (one_le_pow₀ hx)), sub_sub_cancel]
+
+/-- `arcosh` is the right inverse of `cosh` over [1, ∞). -/
+@[simp]
+theorem cosh_arcosh {x : ℝ} (hx : 1 ≤ x) : cosh (arcosh x) = x := by
+  rw [arcosh, cosh_eq, exp_neg, exp_log (by positivity), x_add_sqrt_x_sq_sub_one_inv x hx]
+  ring
+
+@[simp]
+theorem arcosh_eq_zero_iff {x : ℝ} (hx : 1 ≤ x) : arcosh x = 0 ↔ x = 1 := by
+  constructor
+  case mpr => intro hx; subst hx; exact arcosh_zero
+  case mp =>
+  intro h
+  convert congrArg cosh h
+  · rw [cosh_arcosh hx]
+  · rw [cosh_zero]
+
+@[simp]
+theorem sinh_arcosh {x : ℝ} (hx : 1 ≤ x) : sinh (arcosh x) = √(x ^ 2 - 1) := by
+  rw [arcosh, sinh_eq, exp_neg, exp_log (by positivity), x_add_sqrt_x_sq_sub_one_inv x hx]
+  ring
+
+/-- `arcosh` is the left inverse of `cosh` over [0, ∞). -/
+@[simp]
+theorem arcosh_cosh {x : ℝ} (hx : 0 ≤ x) : arcosh (cosh x) = x := by
+rw [arcosh, ← exp_eq_exp, exp_log (by positivity), ← eq_sub_iff_add_eq', exp_sub_cosh,
+    ← sq_eq_sq₀ (sqrt_nonneg _) (sinh_nonneg_iff.mpr hx), ← sinh_sq, sq_sqrt (pow_two_nonneg _)]
+
+@[simp]
+theorem arcosh_nonneg {x : ℝ} (hx : 1 ≤ x) : 0 ≤ arcosh x := by
+  apply log_nonneg
+  calc
+    1 ≤ x := hx
+    _ ≤ x + √(x ^ 2 - 1) := by
+      apply (le_add_iff_nonneg_right _).mpr
+      apply sqrt_nonneg
+
+@[simp]
+theorem arcosh_pos {x : ℝ} (hx : 1 < x) : 0 < arcosh x := by
+  have hx' := le_of_lt hx
+  apply Std.lt_of_le_of_ne (arcosh_nonneg hx')
+  contrapose! hx
+  exact ge_of_eq ((arcosh_eq_zero_iff hx').mp hx.symm).symm
+
+@[simp]
+theorem arcosh_le_arcosh {x y : ℝ} (hx : 1 ≤ x) (hy : 1 ≤ y) : arcosh x ≤ arcosh y ↔ x ≤ y := by
+  rw (occs := .pos [2]) [← cosh_arcosh hx, ← cosh_arcosh hy]
+  rw (occs := .pos [1]) [← abs_of_nonneg (arcosh_nonneg hx), ← abs_of_nonneg (arcosh_nonneg hy)]
+  exact cosh_le_cosh.symm
+
+@[simp]
+theorem arcosh_lt_arcosh {x y : ℝ} (hx : 1 ≤ x) (hy : 1 ≤ y) : arcosh x < arcosh y ↔ x < y := by
+  rw (occs := .pos [2]) [← cosh_arcosh hx, ← cosh_arcosh hy]
+  rw (occs := .pos [1]) [← abs_of_nonneg (arcosh_nonneg hx), ← abs_of_nonneg (arcosh_nonneg hy)]
+  exact cosh_lt_cosh.symm
+
+end Real
+
+/-- Real numbers which are at least 1. -/
+def AOReal := { r : ℝ // 1 ≤ r } deriving
+  Preorder, PartialOrder, SemilatticeInf, SemilatticeSup, DistribLattice,
+  TopologicalSpace, UniformSpace, PseudoMetricSpace,
+  Nontrivial, Inhabited
```

Comment:
I don't think we should have this subtype. It seems that the only reason you're defining it is to serve as the domain / codomain of the restrictions of arcosh / cosh, but that's not a good enough reason. We generally want to avoid upgrading domains or codomains (as sets) to types.

So, instead of building an `Equiv` between the subtypes, I recommend building a `PartialEquiv` instead. Note that this also means I would not have the primed versions `cosh'` and `arcosh'`.